src/shared_realm.cpp

```diff
@@ -267,86 +256,116 @@ bool Realm::reset_file_if_needed(Schema& schema, uint64_t version, std::vector<S
-                    throw InvalidSchemaVersionException(m_schema_version, version);
-                }
-                if (version == m_schema_version) {
-                    ObjectStore::verify_no_changes_required(required_changes);
-                    return true;
-                }
-                return false;
-        }
-        REALM_COMPILER_HINT_UNREACHABLE();
-    };
+Schema Realm::get_full_schema()
+{
+    if (!m_read_only_group)
+        refresh();
+
+    // If the user hasn't specified a schema previously then m_schema is always
+    // the full schema
+    if (m_dynamic_schema)
+        return m_schema;
+
+    // Otherwise we may have a subset of the file's schema, so we need to get
+    // the complete thing to calculate what changes to make
+    if (m_read_only_group)
+        return ObjectStore::schema_from_group(read_group());
 
-    if (no_changes_required())
+    Schema actual_schema;
+    uint64_t actual_version;
+    uint64_t transaction = -1;
+    bool got_cached = m_coordinator->get_cached_schema(actual_schema, actual_version, transaction);
+    if (!got_cached || transaction != m_shared_group->get_version_of_current_transaction().version)
+        return ObjectStore::schema_from_group(read_group());
+    return actual_schema;
+}
+
+void Realm::update_schema(Schema schema, uint64_t version,
+                          MigrationFunction migration_function, bool in_transaction)
+{
+    schema.validate();
+
+    Schema actual_schema = get_full_schema();
+    std::vector<SchemaChange> required_changes = actual_schema.compare(schema);
+
+    if (!schema_change_needs_write_transaction(schema, required_changes, version)) {
+        set_schema(actual_schema, std::move(schema));
         return;
+    }
     // Either the schema version has changed or we need to do non-migration changes
 
     if (!in_transaction) {
-        bool schema_changed = false;
-        m_group->set_schema_change_notification_handler([&]{
-            schema_changed = true;
-            m_schema = ObjectStore::schema_from_group(read_group());
-            m_schema_version = ObjectStore::get_schema_version(read_group());
-            m_coordinator->cache_schema(m_schema, m_schema_version);
-            required_changes = m_schema.compare(schema);
-        });
+        auto old_version = m_shared_group->get_version_of_current_transaction();
         transaction::begin_without_validation(*m_shared_group);
-        add_schema_change_handler();
 
         // Beginning the write transaction may have advanced the version and left
         // us with nothing to do if someone else initialized the schema on disk
-        if (schema_changed && no_changes_required()) {
-            cancel_transaction();
-            return;
+        if (m_new_schema) {
+            actual_schema = std::move(*m_new_schema);
+            m_new_schema = util::none;
+            m_coordinator->cache_schema(actual_schema, m_schema_version,
```

Comment:
It looks like this is doing roughly the same work as `cache_new_schema()`. Is it intentional that we're not updating `m_schema_transaction_version` like `cache_new_schema()` does? Can we find a way to call `cache_new_schema()` here rather than repeating its logic?